Angular value accessor post (#238)
unicorn-utterances/unicorn-utterances: content/blog/angular-components-control-value-accessor/index.md

> It's worth mentioning that unlike the other three methods, this one is entirely optional for implementing a `ControlValueAccessor`. This allows us to disable the component or keep it enabled but is not required for usage with the other methods. `ngModel` and `formControl` will work without this method implemented.

## `registerOnChange` {#register-on-change}

While the previous methods have been implemented in a way that required usage of `markForCheck`, these last two methods are implemented in a bit of a different way. You only need look at the type of the methods on the interface to see as much:

```typescript
registerOnChange(fn: (value: any) => void);
```

As you might be able to deduce from the method type, when `registerOnChange` is called, it passes you a function. You'll then want to store this function in your class instance and call it whenever the user changes data.

```typescript
/** The method to be called to update ngModel */
_controlValueAccessorChangeFn: (value: any) => void = () => {};

registerOnChange(fn: (value: any) => void) {
    this._controlValueAccessorChangeFn = fn;
}
```

> While this code sample shows you how to store the function, but doesn't outline how to call it once stored. You'll want to make sure to call it with the updated value on every update. For example, if you are expecting an `input` to change, you'd want to add it to `(change)` output of the `input`:

Reviewer: weird sentence structure for that first one. 'but doesn't outline how to call it once stored.' is where it reads oddly.

Author: ```suggestion
While this code sample shows you how to store the function, it doesn't outline how to call it once stored. You'll want to make sure to call it with the updated value on every update. For example, if you are expecting an `input` to change, you'd want to add it to `(change)` output of the `input`:

```

```html
<input
    placeholder=""
    [disabled]="disabled"
    [(ngModel)]="value"
    (change)="onChange($event)"
    (blur)="onTouched()"
/>
```

# Consumption {#consume-demo}

Now that we've done that work let's put it all together, apply [the styling from before](#code-demo), and consume the component we've built!

We'll need to start by importing `FormModule` and `ReactiveFormModule` into your `AppModule` for `ngModel` and `formControl` support respectively.

```typescript
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { FormsModule } from '@angular/forms';

import { ReactiveFormsModule } from '@angular/forms';

import { AppComponent } from './app.component';
import { ExampleInputComponent } from './example-input/example-input.component';

@NgModule({
  imports:      [ ReactiveFormsModule, FormsModule, BrowserModule ],
  declarations: [ AppComponent, ExampleInputComponent ],
  bootstrap:    [ AppComponent ]
})
export class AppModule { }
```

Once you have support for them both, you can move onto adding a `formControl` item to your parent component:

```typescript

> import { Component, VERSION } from '@angular/core';

Reviewer: where is VERSION being used?

Author: It's not. I misjudged `VERSION`:

```suggestion
import { Component } from '@angular/core';
```

  modelValue = "";
}
```

Finally, you can pass these options to `ngModel` and `formControl` (or even `formControlName`) and inspect the value directly from the parent itself:

```html
<h1>Form Control</h1>
<app-example-input placeholder="What's your favorite animal?" [formControl]="control"></app-example-input>
<p>The value of the input is: {{control.value}}</p>
<h1>ngModel</h1>
<app-example-input placeholder="What's your favorite animal?" [(ngModel)]="modelValue"></app-example-input>
<p>The value of the input is: {{modelValue}}</p>
```

If done properly, you should see something like this:

<iframe src="https://stackblitz.com/edit/angular-value-accessor-example?ctl=1&embed=1&file=src/app/app.component.ts" style="width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;" sandbox="allow-modals allow-forms allow-popups allow-scripts allow-same-origin"></iframe>

# Form Control Classes

Angular CSS masters might point to [classes that's applied to inputs when various state changes are made](https://angular.io/api/forms/NgControlStatus#css-classes-applied).

These classes include:

- `ng-pristine`
- `ng-dirty`
- `ng-untouched`
- `ng-touched`

> They reflect states so that you can update the visuals in CSS to reflect them. When using `[(ngModel)]`, they won't appear, since nothing is tracking when a component is `pristine` or `dirty`. However, when using `[formControl]` or `[formControlName]`, these classes _will_ appear and act accordingly, thanks to the `registerOnChange` and `registerOnTouched` functions. As such, you're able to display custom CSS logic for when each of these values appears.

Reviewer: `when each of these values appears.` or `when each of these classes appears.`?

Author: ```suggestion
They reflect states so that you can update the visuals in CSS to reflect them. When using `[(ngModel)]`, they won't appear, since nothing is tracking when a component is `pristine` or `dirty`. However, when using `[formControl]` or `[formControlName]`, these classes _will_ appear and act accordingly, thanks to the `registerOnChange` and `registerOnTouched` functions. As such, you're able to display custom CSS logic for when each of these states are met.
```

# Gain Access To Form Control Errors {#form-control-errors}

Something you'll notice that wasn't implemented in the `ControlValueAccessor` implementation is support for checking whether validators are applied. If you're a well-versed Angular Form-ite, you'll recall the ability to [validate forms using validators appended to `FormControl`s](https://angular.io/guide/form-validation). Although a niche situation — since most validation happens at the page level, not the component level — wouldn't it be nice to check when a form is valid or not directly from the component to which the form is attached?

Well, thanks to Angular's DI system, we can do just that!

> However, we'll need to make a few changes to the form input [that we made before](#forwardRef). While we previously implemented a provider for form controls, we now need to manually assign the provider ourselves in the constructor:

Reviewer: ```suggestion
However, we'll need to make a few changes to the form input [we made before](#forwardRef). While we previously implemented a provider for form controls, we now need to manually assign the provider ourselves in the constructor:
```

Reviewer: removing that removes som ambiguity of the order of operations.

@Component({
  selector: "app-example-input",
  templateUrl: "./example-input.component.html",
  styleUrls: ["./example-input.component.css"]
})
export class ExampleInputComponent implements ControlValueAccessor, AfterContentInit {
  constructor(
    @Optional() @Self() public ngControl: NgControl,
    private _changeDetector: ChangeDetectorRef
  ) {
    if (ngControl != null) {
      // Setting the value accessor directly (instead of using
      // the providers) to avoid running into a circular import.
      ngControl.valueAccessor = this;
    }
  }
    
  // ...
}
```

In this code sample, we're using [the `@Self` decorator](https://angular.io/api/core/Self) to tell the dependency injection system that "this component _itself_ should have been provided a `formControl` or `formControlName`". However, we want the component to work even when `FormModule` isn't being used, so we allow the dependency injection to return `null` if nothing's passed by utilizing [the `@Optional` decorator](https://angular.io/api/core/Optional).

Now that you have the `ngControl`, you can access the `formControl` by using `ngControl.control`.

```typescript
ngOnInit() {

>     const control = this.ngControl && this.ngControl.control;

Reviewer: since this is TS, why not use the optional chain?
```suggestion
    const control = this.ngControl?.control;
```

Author: We _could_ use optional chaining, but it's a fairly new feature to Angular (9), so I'd like to keep it as-is

    if (control) {
        console.log("ngOnInit", control);
        // FormControl should be available here
    }
}
```

You have [a ton of different props you're able to access for the control's metadata](https://angular.io/api/forms/NgControl). For example, if you want to check when errors are present, you can do the following:

```typescript
get errors() {

> 	const control = this.ngControl && this.ngControl.control;
>     if (control) {

Reviewer: ```suggestion
    const control = this.ngControl?.control;
    if (control) {
```
alignment and optional chaining

Author: ```suggestion
    const control = this.ngControl && this.ngControl.control;
    if (control) {
```

    	return control.touched && control.errors;
    }
    return null;
}
```

And then reference it in the template:

```html

> <span class="inputLabel" [class.redtext]="errors">{{ placeholder }}</span>

Reviewer: While you can do multiple classes this way, you could also do it with templating
```suggestion
<span class="inputLabel{{errors ? ' redtext' : ''}}">{{ placeholder }}</span>
```

This one is almost entirely personal taste though. I like this method better but a lot of people don't because of the empty string at the end

Author: I'd rather keep things the way we are

```

Now that you have the component implementation, you can add validators to your `FormControl`:

```typescript
import { Component } from '@angular/core';
import {FormControl, Validators} from '@angular/forms';

@Component({
  selector: 'my-app',
  templateUrl: './app.component.html',
  styleUrls: [ './app.component.css' ]
})
export class AppComponent  {
  control = new FormControl('', Validators.required);
}
```

<iframe src="https://stackblitz.com/edit/angular-value-accessor-dep-inject?ctl=1&embed=1&file=src/app/app.component.ts" style="width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;" sandbox="allow-modals allow-forms allow-popups allow-scripts allow-same-origin"></iframe>

Not only do you have [a wide range of Angular-built validators at your disposal](https://angular.io/api/forms/Validators), but you're even able to [make your own validator](https://angular.io/api/forms/Validator)!

# Conclusion {#conclusion}

> Enabling `formControl` and `ngModel` usage is an extremely powerful tool that enables you to have feature-rich and consistent APIs across your form components. Using them, you can ensure that your consumers are provided with the functionality they'd expect in a familiar API to native elements. Hopefully, this article has provided you with more in-depth insight that you're able to take to your own components.

Reviewer: ```suggestion
Enabling `formControl` and `ngModel` usage is an extremely powerful tool that enables you to have feature-rich and consistent APIs across your form components. Using them, you can ensure that your consumers are provided with the functionality they'd expect in a familiar API to native elements. Hopefully, this article has provided you with more in-depth insight that you're able to use with your own components.
```

`to take to` sounds odd to me. If it's just me it's just me, but I wouldn't say that unless I'm looking to physically explain the difference between affect and effect.
